ccl/changefeedccl: TestChangefeedHandlesDrainingNodes failed

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/8731526?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/8731526?buildTab=artifacts#/) on master @ [af3bf237ce3774258a13a0c2aa0e2da22716270f](https://github.com/cockroachdb/cockroach/commits/af3bf237ce3774258a13a0c2aa0e2da22716270f):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/ed07864c133fcb96f5a3b1e88989b58c/logTestChangefeedHandlesDrainingNodes562088760
    test_log_scope.go:79: use -show-logs to present logs inline
    testcluster.go:392: migration-job-find-already-completed: could not register flowID 48111def-d6e9-4bff-8c65-c16be54c648e because the registry is draining
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/ed07864c133fcb96f5a3b1e88989b58c/logTestChangefeedHandlesDrainingNodes562088760
--- FAIL: TestChangefeedHandlesDrainingNodes (3.42s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-24582

Epic CRDB-11732

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8944163?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8944163?buildTab=artifacts#/) on master @ [71d8239e3a06c0d3abbb6ea18f0d657ef659a812](https://github.com/cockroachdb/cockroach/commits/71d8239e3a06c0d3abbb6ea18f0d657ef659a812):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/f236cfce7e32f4b492de4b4f1cd80c96/logTestChangefeedHandlesDrainingNodes975755735
    test_log_scope.go:79: use -show-logs to present logs inline
    testcluster.go:394: migration-job-find-already-completed: could not register flowID d6145562-bf38-4bef-a227-019df683ce8c because the registry is draining
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f236cfce7e32f4b492de4b4f1cd80c96/logTestChangefeedHandlesDrainingNodes975755735
I230306 21:05:39.976286 41057 2@rpc/context.go:2473  [T1,n2,rnode=1,raddr=127.0.0.1:36073,class=rangefeed,rpc] 319  connection is now ready
I230306 21:05:39.978357 41105 2@rpc/context.go:2473  [T1,n1,rnode=2,raddr=127.0.0.1:40705,class=default,rpc] 320  connection is now ready
--- FAIL: TestChangefeedHandlesDrainingNodes (1.30s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8987919?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8987919?buildTab=artifacts#/) on master @ [b47dba8644749af91e966e1d9277f27aa33a9852](https://github.com/cockroachdb/cockroach/commits/b47dba8644749af91e966e1d9277f27aa33a9852):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/f236cfce7e32f4b492de4b4f1cd80c96/logTestChangefeedHandlesDrainingNodes778988401
    test_log_scope.go:79: use -show-logs to present logs inline
    testcluster.go:394: migration-job-find-already-completed: could not register flowID abe84fa1-0207-49d2-8e90-4249b33e8768 because the registry is draining
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f236cfce7e32f4b492de4b4f1cd80c96/logTestChangefeedHandlesDrainingNodes778988401
--- FAIL: TestChangefeedHandlesDrainingNodes (2.08s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -5892,15 +5892,15 @@ func TestChangefeedHandlesDrainingNodes(t *testing.T) {
 
 	skip.UnderRace(t, "Takes too long with race enabled")
 
-	shouldDrain := true
+	var shouldDrain int32
 	knobs := base.TestingKnobs{
 		DistSQL: &execinfra.TestingKnobs{
 			DrainFast:  true,
 			Changefeed: &TestingKnobs{},
 			Flowinfra: &flowinfra.TestingKnobs{
 				FlowRegistryDraining: func() bool {
-					if shouldDrain {
-						shouldDrain = false
+					if atomic.LoadInt32(&shouldDrain) > 0 {
+						atomic.StoreInt32(&shouldDrain, 0)
 						return true
 					}
 					return false
@@ -5948,6 +5948,7 @@ func TestChangefeedHandlesDrainingNodes(t *testing.T) {
 	f, closeSink := makeFeedFactory(t, randomSinkType(feedTestEnterpriseSinks), tc.Server(1), tc.ServerConn(0))
 	defer closeSink()
 
+	atomic.StoreInt32(&shouldDrain, 1)
 	feed := feed(t, f, "CREATE CHANGEFEED FOR foo")
 	defer closeFeed(t, feed)
 

